pandas/_libs/lib.pyx

```diff
@@ -2315,3 +2312,199 @@ def fast_multiget(dict mapping, ndarray keys, default=np.nan):
+cdef cnp.ndarray[object] _concat_date_cols_sequence(tuple date_cols,
+                                                    Py_ssize_t rows_count,
+                                                    Py_ssize_t col_count,
+                                                    bint keep_trivial_numbers):
+    """
+    Concatenates `rows_count` elements from each `col_count` sequences
+    in `date_cols` into strings.
+
+    Parameters
+    ----------
+    date_cols : tuple of sequences
+    rows_count : Py_ssize_t
+    col_count : Py_ssize_t
+    keep_trivial_numbers : bool, default False
+        if True and len(date_cols) == 1, then
+        conversion (to string from integer/float zero) is not performed
+
+    Returns
+    -------
+    arr_of_rows : ndarray (dtype=object)
+    """
+    cdef:
+        Py_ssize_t col_idx, row_idx
+        list list_to_join
+        cnp.ndarray[object] result
+        object[:] result_view
+
+    result = np.zeros(rows_count, dtype=object)
+    # create memoryview of result ndarray - more effecient indexing
+    result_view = result
+
+    if col_count == 1:
+        for row_idx, item in enumerate(date_cols[0]):
+            result_view[row_idx] = convert_to_unicode(item,
+                                                      keep_trivial_numbers)
+    else:
+        list_to_join = [None] * col_count
+        for row_idx in range(rows_count):
+            for col_idx, array in enumerate(date_cols):
+                list_to_join[col_idx] = convert_to_unicode(array[row_idx],
+                                                           False)
+            result_view[row_idx] = PyUnicode_Join(' ', list_to_join)
+
+    return result
+
+
+def _concat_date_cols(tuple date_cols, bint keep_trivial_numbers=True):
+    """
+    Concatenates elements from sequences in `date_cols` into strings.
+
+    Parameters
+    ----------
+    date_cols : tuple of sequences
+    keep_trivial_numbers : bool, default True
+        if True and len(date_cols) == 1, then
+        conversion (to string from integer/float zero) is not performed
+
+    Returns
+    -------
+    arr_of_rows : ndarray (dtype=object)
+
+    Examples
+    --------
+    >>> dates=np.array(['3/31/2019', '4/31/2019'], dtype=object)
+    >>> times=np.array(['11:20', '10:45'], dtype=object)
+    >>> result = _concat_date_cols((dates, times))
+    >>> result
+    array(['3/31/2019 11:20', '4/31/2019 10:45'], dtype=object)
+    """
+    cdef:
+        Py_ssize_t rows_count = 0, col_count = len(date_cols)
+
+    if col_count == 0:
+        return np.zeros(0, dtype=object)
+
+    rows_count = min(len(array) for array in date_cols)
+
+    if all(util.is_array(array) for array in date_cols):
+        # call specialized function to increase performance
+        return _concat_date_cols_numpy(date_cols, rows_count, col_count,
+                                       keep_trivial_numbers)
+    else:
+        return _concat_date_cols_sequence(date_cols, rows_count, col_count,
```

Comment:
how is this even possible to hit code path? (or alt the other one); meaning aren't these *always* a sequence or already a numpy array? how can they ever be mixed? it seems adding a lot of code for cases which don't exist.

Reply:
`date_cols` is (actually a variable-length) tuple coming from outside as a parameter which contains a list of columns to concatenate.
Now what this piece of code does is it checks if _all_ columns are Numpy arrays, and, if they are, it uses specialized faster version. If any of columns isn't a Numpy array it falls back to general, slower implementation.

That said, we're actually _not_ expecting a user to feed in a mix of Numpy arrays and lists (but a user still can :) ), what we expect is that a user might want to concatenate a few `list`-s or `tuple`-s instead of a few Numpy arrays.